docs/proposals/dockerfile_component_detection.md

```diff
@@ -0,0 +1,58 @@
+# Enhance port detection for dockerfiles that are not in a component.
+
+## Related GitHub Issues
+* https://github.com/redhat-developer/alizer/issues/235
+
+## Background
+
+What is a component?
+
+Components are defined in the `languages-customization.yml` as a language that has specfic `configuration_files`.
+
+Example of a Go component:
+
+```yaml
+Go:
+  exclude_folders:
+    - "vendor"
+  configuration_files:
+    - "go.mod"
+  component: true
+```
+
+What happens when the repo being scanned has component(s) found but also contains a dockerfile one level outside of the repo?
+
+```shell
+.repo-being-detected-by-alizer
+├── component/
+│   └── dockerfile
+└── dockerfile
+```
+
+The dockerfile in the root level would not be used during analysis because it does not have a component definition in the `languages-customization.yml`
+
+## Potential Solution
+
+Introduce a new component field called `orphan_component` and add it to the `Dockerfile` language:
```

Comment:
Why do we call it orphan_component instead of Dockerfile component?

Reply:
In the example above, what is the priority of the search order? If there is a Dockerfile at the root, do we consider that as a Dockerfile component?

Reply:
We decided not to use "docker_component" or anything that is closely coupled to Dockerfiles so we can resuse the field if there were ever another "orphaned" component (as in we wanted to define a component like Dockerfile where there isn't really a language or framework tied to it.)

We also didn't want to use the existing field `component` as it doesn't really match the others (i.e. C#, F#, Go, Java, etc) because they're languages, compared to a single Dockerfile.

Also open to renaming the field too if there's something more accurate

Reply:
Do you mean Dockerfile support will be handled differently in those two cases? Do we still try to do port detection for Dockerfile on root? It is still not clear to me why those two dockerfile cases will need to be processed differently.

Reply:
Currently, port detection only happens on found components. Because the Dockerfile on root isn't a component, it doesn't do port detection in that case.

So the proposal is to turn the Dockerfile on the root into an orphan_component, and then apply the same port detection logic against it

Reply:
> We decided not to use "docker_component" or anything that is closely coupled to Dockerfiles so we can resuse the field if there were ever another "orphaned" component (as in we wanted to define a component like Dockerfile where there isn't really a language or framework tied to it.)

I am not sure if orphan_component will be intuitive enough for people who are new to Alizer and understand the processing logic. If this type of component is to cover all non-language-specific cases, have we considered calling it something like "nonlanguage_component" to make it clear?

Reply:
+1 for `nonlanguage_component`

Reply:
or `container_component`?  Since Dockerfile is part of language.yaml, Linguist considers it a [language](https://github.com/github-linguist/linguist/blob/master/lib/linguist/languages.yml#L1540C1-L1554).  To call it anonlanguage would be confusing in this context.   I know you mentioned you wanted to make this generic enough to future proof this but if we ever do encounter some other use case, maybe we can just add a new component classification and treat it in the same way as `container_component`?

Reply:
That is a good point, I can't really think of another use case and creating a new component classification wouldn't be hard since each enricher is language specific

If we did want to stick with a generic name, maybe we could use `domain_specific_component`? Docker does consider the [Dockerfile syntax to be a DSL](https://docs.docker.com/build/guide/intro/#:~:text=A%20Dockerfile%20is%20a%20text,language%2C%20called%20the%20Dockerfile%20syntax)

Reply:
As long as we are confident enough that DSLs are always going to be processed this way: https://en.wikipedia.org/wiki/Domain-specific_language

Reply:
Settled on `container_component` with the intent that new component classifications can be created in the future if needed